Sources/Holocron/StorageProvider.swift

```diff
@@ -25,31 +25,34 @@ private func performAsynchronously<T>(_ block: @escaping () throws -> T, complet
 // MARK: StorageProvider
 /// An interface for a simple key value store.
 public protocol StorageProvider {
+    /// The type to use for keys. In most cases, this will be String, but at the very least, it should be opaque.
+    associatedtype Key
```

Comment:
Does this need to be an associated type? Could it simply be a struct that happens to be `RawRepresentable`? That should simplify the protocol requirements and make the usage of the `Key` completely additive - meaning in theory it could be used with all `StorageProvider` types.

Something like: 

```swift
public struct Key: Hashable, RawRepresentable, ExpressibleByStringLiteral {
    public var rawValue: String
    
    public init(rawValue: String) {
        self.rawValue = rawValue
    }
    
    public init(stringLiteral value: String) {
        self.rawValue = value
    }
}
```

Reply:
We could definitely do this, but I'm not seeing the advantage here vs. `typealias Key = String` if we're defining the Key type ourselves. I can see the benefit of letting users define Key however they want, but if we're locking them into a specific Key type, it should provide some advantage over using a String, right?

Reply:
I also think that we should loosen the requirements for `Key`. Requiring it to be `ExpressibleByStringLiteral` and `RawRepresentable` doesn't give users much additional flexibility or safety vs. using `String`. I think we should either require that it be `Equatable`, or leave the requirements empty and include a note that says it should be opaque and work the way you'd expect a `Key` to.

Reply:
Well for me, the type safety is the advantage. Having a compiler checked variable versus a String is a huge advantage. It could also eliminate the possible `XKey``YKey` and `ZKey` interface pollution.

```swift 
provider.write(myObject, to: .someComplicatedKeyName)
```

vs 

```swift 
provider.write(myObject, to: "someComplciatedKeyName")
```

It also lets you keep the `StorageProvider` protocol from having an associated type, meaning it can be used in a more canonical way.

Reply:
That makes sense. I didn't consider the enum-style key names, and we'd need a custom key type to prevent polluting the `String` "namespace". However, if we type all of our keys as that custom `Key` type, we lock users out of using strings if they want to.

So what if we create the `Key` type as you suggested, but we also make it `CustomStringConvertible`? We could then type all of the APIs to take `CustomStringConvertible` values as keys, which frees up those who want to use raw strings to do so. If we don't want to muddy the use of `CustomStringConvertible`, we could also make our own protocol and extend `String` to conform to it. If a user is importing a `StorageProvider`, she'll also be importing the custom protocol + String conformance.

Reply:
Well if `Key` is `ExpressibleByStringLiteral` you can use a `String` and it gets implicitly converted into a `Key`.

Reply:
@wmcginty I guess I forgot what `ExpressibleByStringLiteral` means...those changes have landed in my branch.

Reply:
@pranjalsatija Looking good!